### PR TITLE
device: make Peer.device an atomic.Value

### PR DESCRIPTION
**Pull request by josharian**
The removal of the RLock/RUnlock in Peer.timersActive provides a 10% throughput improvement for Tailscale communications.

This is going the wrong direction; we're trying to get closer to upstream wireguard-go. Note that upstream wireguard-go doesn't have this lock in the first place. I don't know offhand whether this is a bug upstream or whether our other changes necessitated this lock.

As a result, I'm not sure whether we should merge this or not.

### CODE REVIEW

**Comment by josharian** [comment]
OK, upstream wireguard is missing this as a consequence of not having 7d52df8aa477716c4d39498143b7fc3d306d2d8b, which is a fix for data races which upstream has. So I guess we should consider merging this, and later rebasing it down as part of upstreaming the data race fixes. I may add more such crossover perf/data race fixes in this PR, as I find them...

**Comment by bradfitz** [comment]
IIRC a number of our data race fixes were for races that don't/can't manifest upstream because they only came about due to our fork's wgcfg additions or other hooks we added violating perhaps undocumented invariants. @crawshaw knows such history.

File: device/peer.go [hunk 1]
```diff
@@ -77,6 +77,14 @@ type Peer struct {
 	cookieGenerator CookieGenerator
 }
 
+func (p *Peer) device() *Device {
+	v := p.deviceValue.Load()
```
Comment (bradfitz) [inline device/peer.go:81]: I like to write

    V, _ := x.Load().(*T)
   Return v

File: device/peer.go [hunk 2]
```diff
@@ -104,7 +112,7 @@ func (device *Device) NewPeer(pk wgcfg.Key) (*Peer, error) {
 	defer peer.Unlock()
 
 	peer.cookieGenerator.Init(pk)
-	peer.device = device
+	peer.deviceValue.Store(device)
```
Comment (bradfitz) [inline device/peer.go:115]: why do I only see one Store in this entire change? If there's only one Store and the Peer is allocated just a line or two up on line 110 here, why don't we just access the `.device` field directly without the `peer.RLock()` call at all in `device/timers.go`?

**Comment by crawshaw** [comment]
I was chatting with Jason recently about this and he thinks these variables should be configurable on running peers using the UAPI protocol, which means the race exists upstream. At the time I did this work I couldn't find the code to trigger that, it may be the UAPI protocol for the kernel wireguard is ahead of wireguard-go (or it could be I just missed it).

So either this data race is relevant to upstream, or it would be if UAPI was extended to match kernel wireguard.

**Comment by bradfitz** [comment]
So what's the idea? You can atomically move a peer between network devices at runtime with kernel wireguard?

**Pull request closed without merging**